### PR TITLE
Add filters and disable logic to magic item selection

### PR DESCRIPTION
**Pull request by korsvanloon**
- A unit does not see magic items anymore that are higher than its allowed maximum.
- All items that are above the currently available left over points are disabled
- All items of a currently selected type (except stackable) are disabled.
- Added a maximum for the stackable input based on the points remaining (or an existing configured maximum)
- The radio variant (and the no-banner option) is removed since we now have stricter limitation rules.
- Add some simple styling for a disabled checkbox state.

Scherm­afbeelding 2024-01-24 om 07 51 49
Scherm­afbeelding 2024-01-24 om 07 55 51

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,6 +36,9 @@
     "workbox-strategies": "^5.1.4",
     "workbox-streams": "^5.1.4"
   },
+  "devDependencies": {
+    "eslint": "^7.11.0"
```
Comment (nthiebes) [inline package.json:40]: I have prettier set-up, so not sure if it's needed.

File: src/App.css [hunk 2]
```diff
@@ -215,6 +215,11 @@ hr {
   width: 100%;
 }
 
+.radio__input:disabled + .radio__label,
+.checkbox__input:disabled + .checkbox__label {
+  opacity: 0.3;
```
Comment (nthiebes) [inline src/App.css:220]: You can also add `cursor: default;`

File: src/pages/magic/Magic.js [hunk 3]
```diff
@@ -268,29 +268,21 @@ export const Magic = ({ isMobile }) => {
...
-    }
+  const getCheckbox = ({
+    isChecked,
+    selectedAmount,
+    magicItem,
+    itemGroup,
+    isConditional,
+    isTypeLimitReached,
+  }) => {
+    const isCommand = Boolean(unit?.command[command]);
```
Comment (nthiebes) [inline src/pages/magic/Magic.js:279]: This also needs to always check for `?.magic?.maxpoints` otherwise the page will crash for some units. Some units have empty command values `[]`, some have the magic key but no data. Therefore, checking for maxPoints is necessary. I had a lot of crashes on the magic page before because of a mix between command magic items and general items.

**Review by nthiebes — COMMENTED** [review]
First of all: Thanks for supporting and putting in the effort! 🎉

_A unit does not see magic items anymore that are higher than its allowed maximum._
That's a great improvement! 💪

_All items that are above the currently available left over points are disabled_
I kinda like it, but I also have doubts since it requires an additional click if you want to switch items 🤔. It is to some degree already covered by the error messages/icons.

_All items of a currently selected type (except stackable) are disabled._
1. This is basically what a radio input field is meant for. However, I do agree that the "No banner" option isn't nice. There are a few options here: Disable other checkboxes, radio fields with "No banner/item" (we could filter those so they don't show up on a unit card) or a clear button (don't like it).
2. Some item categories like the dwarven runes need other options to still be available. So only when `mutuallyExclusive` is true. Maybe other factions also have items that are not mutually exclusive.

_Added a maximum for the stackable input based on the points remaining (or an existing configured maximum)_
The numbers input doesn't appear for me at all for stackable items?

_The radio variant (and the no-banner option) is removed since we now have stricter limitation rules._
See previous answer.

_Add some simple styling for a disabled checkbox state._
Sweet! 👍

**Comment by nthiebes** [comment]
Selecting a dwarven rune disables all master runes. Should be that selecting a master rune deselects all master runes of that item category. That would be sweet. Also the max of three runes should only apply to each item category.

File: package.json (hunk 1 above)
Comment (korsvanloon) [reply package.json:40]: I think you've installed it globally on your machine.
I couldn't start the project with `yarn start` because it was complaining eslint was missing.
Also since I (and many other developers) run many projects with many different eslint versions, I cannot install this globally :)

**Comment by korsvanloon** [comment]
> _All items that are above the currently available left over points are disabled_ 
>
> I kinda like it, but I also have doubts since it requires an additional click if you want to switch items 🤔. It is to some degree already covered by the error messages/icons.
>
> _All items of a currently selected type (except stackable) are disabled._
> 
> 1. This is basically what a radio input field is meant for. However, I do agree that the "No banner" option isn't nice. There are a few options here: Disable other checkboxes, radio fields with "No banner/item" (we could filter those so they don't show up on a unit card) or a clear button (don't like it).
> 2. Some item categories like the dwarven runes need other options to still be available. So only when `mutuallyExclusive` is true. Maybe other factions also have items that are not mutually exclusive.

I tried to find a unified solution that could deal with the semi-radio behavior of magic items. So the way the UX works for all magic items is the same.
Radio buttons obviously does not work for many situations even though you most of the time can only select 1 of a list.

So that's why I settled for just making everything a checkbox and disable everything that's not allowed.

Regarding the extra clicks:
You already needed additional clicks in a checkbox list if you wan to switch items. In the previous solution, you'd exceed the points first and then have to deselect the old one. In the new solution you first deselect an option and than see what now is available. It's a bit different, but very clear IMO. For radio lists this is not the case anymore of course, but the only radio list I could see was magic banners.

Indeed the the error messages and icons redundant now.

Thanks for bringing up Dwarven runes. Totally forgot to check those.

> _Added a maximum for the stackable input based on the points remaining (or an existing configured maximum)_ 
> The numbers input doesn't appear for me at all for stackable items?

It should if you _can_ fill in more than 1. If you can only fill in 1 (because you don't have enough points), the input will not show up. If that's not the case, I made a last minute bug :P


So all in all, do you agree with my disabled checkbox approach?
I can fix the logic for the Dwarven runes and if you want I can also remove the error messages/icons which are not necessary anymore I believe (is that correct?).

**Comment by nthiebes** [comment]
_It should if you can fill in more than 1. If you can only fill in 1 (because you don't have enough points), the input will not show up. If that's not the case, I made a last minute bug :P_
It didn't work for any common item even at low points 🤷‍♂️.

_So all in all, do you agree with my disabled checkbox approach?_
I have to say I like how it feels, it's more intelligent and moves the responsibility of making mistakes from the users to the app/devs. That's also why we need to do a lot of testing for all the edge cases. Dwarven runes, outdated data structure, and all the many "magic" categories like chaos mutations, knighlty honours etc for all the factions. But I think if we do that it will be pretty sweet! Thanks again for your initiative, that's super helpful! 🎉

_I can fix the logic for the Dwarven runes and if you want I can also remove the error messages/icons which are not necessary anymore I believe (is that correct?)._
Let's go! Should be the icons in the magic header and on the units page.

I'm right now working on detachments and managing all the contributions. So for now I can't support much besides testing.

**Comment by korsvanloon** [comment]
@nthiebes I applied your feedback and improved the logic for Dwarfen runes.

Note that the actual rules for runes are a bit more complex.
I only check that you cannot have more than 3 runes of the same type. But other rules such as no more than 3 total runes per category etc. I do not check (yet).
Same goes for uniqueness of magic items or master runes within your list. (multiple characters can have the same magic item still).
But my main goal is to at least not wrongly restrict the user.

I also added some simple unit tests for magic item limitation rulings. So if we would like to expand magic item restriction logic, we can easily see what's supported and test if changes don't break other rulings.

Tomorrow I can test the Bretonnia magic.
Can you help me out with testing the Forces of Evil part? I do not possess that book yet.

**Comment by nthiebes** [comment]
> @nthiebes I applied your feedback and improved the logic for Dwarfen runes.
Sounds great! I'll take a closer look this evening.

> Also I only show the number input for arcane, enchanted items and runes. Since it doesn't make sense to have 2 of the same magical swords or sets of armor.
Common magic items can be used multiple times per category, army and per character with no limit besides the max magic item points. I just confirmed that in the rulebook.

> Note that the actual rules for runes are a bit more complex. I only check that you cannot have more than 3 runes of the same type. But other rules such as no more than 3 total runes per category etc. I do not check (yet). Same goes for uniqueness of magic items or master runes within your list. (multiple characters can have the same magic item still). But my main goal is to at least not wrongly restrict the user.
Yeah, dwarfven runes are complex. But what we have now is a good start 👍

> I also added some simple unit tests for magic item limitation rulings. So if we would like to expand magic item restriction logic, we can easily see what's supported and test if changes don't break other rulings.
Sweet! Something I did not get to yet. The last weeks since the preorder started were crazy. I have yet to properly read the rules or plan my army...

> Tomorrow I can test the Bretonnia magic. Can you help me out with testing the Forces of Evil part? I do not possess that book yet.
Are you part of the Discord? PM me there and I'll send it to you.

**Comment by korsvanloon** [comment]
> Common magic items can be used multiple times per category, army and per character with no limit besides the max magic item points. I just confirmed that in the rulebook.

You are right, but the rules also say you can only use 1 sword and 1 set of armour etc at a time.
Why then allow 1 character to have 2 Sword of Mights? Even though it is literally possible, I thought removing the inputs at those instances makes sense. I might oversee an edge case though.

**Comment by nthiebes** [comment]
For non-common items, it's true that they can only exist once, but for the common ones it's pretty clear that there's no limitation.

**Comment by korsvanloon** [comment]
I updated the PR to allow all common items - even weapons, armor and banners - to be taken multiple times by 1 character.

I double checked all the other magic items (virtues, marks, spites) and nothing really special was going on there.

Maybe good to know regarding how Dwarven runes are modelled now:
- all runes except the master runes are stackable
- if a rune has an * in the name, it's a common rune and it has a maximum of 3 and otherwise 1

A stackable rune with a maximum of 1 means that you can only take it once per character, but can combine it with other runes or the same type.


I hope you agree this PR is now ready for merging :)

File: src/pages/magic/Magic.js [hunk 4]
```diff
@@ -268,29 +272,20 @@ export const Magic = ({ isMobile }) => {
...
-    }
+  const getCheckbox = ({
+    isChecked,
+    selectedAmount,
+    magicItem,
+    itemGroup,
+    isConditional,
+    isTypeLimitReached,
+  }) => {
+    const isCommand = Boolean(unit?.command[command]?.magic?.maxpoints);
```
Comment (nthiebes) [inline src/pages/magic/Magic.js:283]: I have changed the commands magic section so they can now also have magic items with no points limit. It means we should check for `?.magic?.types.length`.

**Review by nthiebes — COMMENTED** [review]
Super awesome! 🙌
I can only properly test it on Sunday since I'm busy tomorrow (I also have my first game 🥳). But it looks excellent, and it's also nice to see some first tests 😄.

**Comment by nthiebes** [comment]
Alright, I merged your PR with a few changes.

- Selecting a unique magic item disabled all other except common ones of that category - that was correct. However, selecting a common item first disabled all others. For fixing that, I added `!item.stackable` to the `isTypeLimitReached` variable.
- I took a closer look at the runes logic again and a few adjustments were needed. I added a general limit of 3 runes per item category. Only the common runes marked with * are stackable, so I removed that option for the other runes. Master runes can only exist once per item and army (not impemented yet), but they can still be combined with other runes up to a max of 3 per item. That differs from other magic items, so it needed its own logic.

The changes are in the main branch so you can also take a look.

Thanks again for contributing this great feature! 🥳 I'll do a little more testing and then deploy.
